File: web_version/pmugg_files/shapes3D/bspPlane.js
ms.bspPlane = function(n, d) {
	this.n = n;
	this.d = d;
};

ms.bspPlane.create = function(face) {
	var endpoint = face.getEndpoints()[0];
	var n = endpoint.faceType().getNormal();
	var d = n.dot(endpoint.getPosition());
	return new ms.bspPlane(n, d);
};

ms.bspPlane.parallelEps = 1e-3;
ms.bspPlane.coPlanarEps = 1e-5;

ms.bspPlane.prototype.getN = function() {
	return this.n;
};

ms.bspPlane.prototype.getD = function() {
	return this.d;
};

ms.bspPlane.prototype.isParallel = function(planeB) {
	return Math.abs(this.n.dot(planeB.getN())) > 1 - ms.bspPlane.parallelEps;
};

ms.bspPlane.prototype.sameD = function(planeB) {
	return Math.abs(this.d - planeB.d) < ms.bspPlane.coPlanarEps;
};

// Determine if a query point is above (+1), below (-1), or on the plane (0).
ms.bspPlane.prototype.sign = function(query) {
	var d = this.n.dot(query);
	if (d > this.d + ms.bspPlane.coPlanarEps) {
		return 1;
	} else if (d > this.d - ms.bspPlane.coPlanarEps) {
		return 0;
	} else {
		return -1;
	}
};

ms.bspPlane.prototype.crossingPoint = function(p0, p1) {
	var d0 = this.n.dot(p0);
	var d1 = this.n.dot(p1);
	var s = (this.d - d0) / (d1 - d0);
	return ms.vec3.lerp(p0, p1, s);
};

File: cpp_version/pmugg/shapes3D/bsp_plane.cpp
#include "bsp_plane.h"
#include "face.h"
#include "face_type.h"
#include "endpoint.h"

namespace ms {

BspPlane::BspPlane(const Vec3& n, float dist)
    : normal(n)
    , d(dist) {}

BspPlane BspPlane::create(Face* face) {
    auto* endpoint = face->getEndpoints()[0];
    Vec3 n = endpoint->getFaceType()->getNormal();
    float d = n.dot(endpoint->getPosition());
    return BspPlane(n, d);
}

bool BspPlane::isParallel(const BspPlane& planeB) const {
    return std::abs(normal.dot(planeB.getNormal())) > 1.0f - PARALLEL_EPS;
}

bool BspPlane::sameD(const BspPlane& planeB) const {
    return std::abs(d - planeB.d) < COPLANAR_EPS;
}

int BspPlane::sign(const Vec3& query) const {
    float dist = normal.dot(query);
    if (dist > d + COPLANAR_EPS) {
        return 1;
    } else if (dist > d - COPLANAR_EPS) {
        return 0;
    } else {
        return -1;
    }
}

Vec3 BspPlane::crossingPoint(const Vec3& p0, const Vec3& p1) const {
    float d0 = normal.dot(p0);
    float d1 = normal.dot(p1);
    float s = (d - d0) / (d1 - d0);
    return Vec3::lerp(p0, p1, s);
}

} // namespace ms 